### PR TITLE
feat(app, step-generation): visualize the step timeline and deck state of a protocol

### PR DESCRIPTION
**Pull request by b-cooper**
# Overview

Create an experimental tool, behind a feature flag, that allows users to inspect the timeline of
atomic commands and the resulting deck state of the robot at ever frame of the timeline.

NOTE: This is for Flex only and also does not yet support the 96-channel

Demo:

https://github.com/Opentrons/opentrons/assets/66035149/e9822fe3-6feb-46ec-a185-a9946a730ecf

co-authored by @jerader 

Turn on the Protocol timeline feature flag and click on the "go to timeline" button from the protocol overflow menu of a Flex protocol with no 96-channel.

# Risk assessment
low, behind a feature flag

### CODE REVIEW

File: app/src/organisms/ProtocolDetails/AnnotatedSteps.tsx [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+  ProtocolAnalysisOutput,
+  RunTimeCommand,
+} from '@opentrons/shared-data'
+
+interface AnnotatedStepsProps {
+  analysis: CompletedProtocolAnalysis | ProtocolAnalysisOutput
+  currentCommandIndex?: number
+}
+
+export const AnnotatedSteps = (props: AnnotatedStepsProps): JSX.Element => {
```
Comment (koji) [inline app/src/organisms/ProtocolDetails/AnnotatedSteps.tsx]:
```suggestion
export function AnnotatedSteps(props: AnnotatedStepsProps): JSX.Element {
```

File: step-generation/src/commandCreators/atomic/comment.ts [hunk 2]
```diff
@@ -0,0 +1,24 @@
+import { uuid } from '../../utils'
+import type { CommentCreateCommand } from '@opentrons/shared-data'
+import type { CommandCreator } from '../../types'
+
+export const comment: CommandCreator<CommentCreateCommand['params']> = (
```
Comment (jerader) [inline step-generation/src/commandCreators/atomic/comment.ts:5]: this isn't actually wired up anywhere but will probably be used in the future. Lots of python protocols use comment commands and we want to implement similar behavior in PD at some point

File: app/src/organisms/ProtocolDetails/AnnotatedSteps.tsx [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+    }
+  `
+
+  return (
+    <Flex
+      css={HIDE_SCROLLBAR}
+      flexDirection={DIRECTION_COLUMN}
+      maxHeight="82vh"
+      flex="1 1 0"
+      overflowY="auto"
```
Comment (koji) [inline app/src/organisms/ProtocolDetails/AnnotatedSteps.tsx]:
```suggestion
      overflowY={OVERFLOW_AUTO}
```

File: app/src/organisms/ProtocolTimelineScrubber/index.tsx [hunk 4]
```diff
@@ -0,0 +1,608 @@
...
+  RobotType,
+  RunTimeCommand,
+} from '@opentrons/shared-data'
+import type {
+  LocationLiquidState,
+  ModuleTemporalProperties,
+  PipetteEntity,
+  TimelineFrame,
+} from '@opentrons/step-generation'
+import type { ViewportListRef } from 'react-viewport-list'
```
Comment (koji) [inline app/src/organisms/ProtocolTimelineScrubber/index.tsx]:
```suggestion
import type { ViewportListRef } from 'react-viewport-list'
import type {
  CompletedProtocolAnalysis,
  LabwareLocation,
  ProtocolAnalysisOutput,
  RobotType,
  RunTimeCommand,
} from '@opentrons/shared-data'
import type {
  LocationLiquidState,
  ModuleTemporalProperties,
  PipetteEntity,
  TimelineFrame,
} from '@opentrons/step-generation'
```

File: app/src/organisms/ProtocolTimelineScrubber/index.tsx [hunk 5]
```diff
@@ -0,0 +1,608 @@
...
+    0
+  )
+  const [isPlaying, setIsPlaying] = React.useState<boolean>(true)
+
+  const currentCommandsSlice = commands.slice(0, currentCommandIndex + 1)
+  const { frame, invariantContext } = getResultingTimelineFrameFromRunCommands(
+    currentCommandsSlice
+  )
+  const handlePlayPause = (): void => {
+    setIsPlaying(!isPlaying)
```
Comment (koji) [inline app/src/organisms/ProtocolTimelineScrubber/index.tsx:81]:
```suggestion
    setIsPlaying(prevIsPlaying => !prevIsPlaying)
```

File: app/src/organisms/ProtocolTimelineScrubber/index.tsx [hunk 6]
```diff
@@ -0,0 +1,608 @@
...
+
+  React.useEffect(() => {
+    if (isPlaying) {
+      const intervalId = setInterval(() => {
+        setCurrentCommandIndex(prev => {
+          const nextIndex = prev < commands.length - 1 ? prev + 1 : 0
+          commandListRef.current?.scrollToIndex(nextIndex)
+          return nextIndex
+        })
+      }, 1000)
```
Comment (koji) [inline app/src/organisms/ProtocolTimelineScrubber/index.tsx]: nit 
make a const for this?

```suggestion
      }, 1000)
```

File: app/src/organisms/ProtocolTimelineScrubber/index.tsx [hunk 7]
```diff
@@ -0,0 +1,608 @@
...
+  )
+
+  return (
+    <Flex
+      height="95vh"
+      flexDirection={DIRECTION_COLUMN}
+      gridGap={SPACING.spacing8}
+    >
+      <Flex gridGap={SPACING.spacing8} flex="1 1 0">
+        <Flex flex="1 1 0" width="300px" height="70vh">
```
Comment (koji) [inline app/src/organisms/ProtocolTimelineScrubber/index.tsx]:
```suggestion
        <Flex flex="1 1 0" width="18.75rem" height="70vh">
```

File: app/src/organisms/ProtocolTimelineScrubber/index.tsx
Comment (koji) [inline app/src/organisms/ProtocolTimelineScrubber/index.tsx:1]: probably exporting functions (PipetteMountViz etc) would make things easy. (testing and maintaining)

File: app/src/organisms/ProtocolTimelineScrubber/index.tsx [hunk 8]
```diff
@@ -0,0 +1,608 @@
...
+          pipetteId={rightPipetteId}
+          pipetteEntity={rightPipetteEntity}
+          timelineFrame={frame.robotState}
+          analysis={analysis}
+        />
+      </Flex>
+      <Flex
+        ref={wrapperRef}
+        alignSelf={ALIGN_STRETCH}
+        overflowY="scroll"
```
Comment (koji) [inline app/src/organisms/ProtocolTimelineScrubber/index.tsx]:
```suggestion
        overflowY={OVERFLOW_SCROLL}
```

File: app/src/organisms/ProtocolTimelineScrubber/index.tsx [hunk 9]
```diff
@@ -0,0 +1,608 @@
...
+      />
+      <Flex alignSelf={ALIGN_STRETCH} justifyContent={JUSTIFY_SPACE_BETWEEN}>
+        <LegacyStyledText as="p">1</LegacyStyledText>
+        <LegacyStyledText as="p">{commands.length}</LegacyStyledText>
+      </Flex>
+      {currentCommandIndex !== 0 &&
+      currentCommandIndex !== commands.length - 1 ? (
+        <LegacyStyledText
+          as="p"
+          marginLeft={(currentCommandIndex / (commands.length - 1)) * 886}
```
Comment (koji) [inline app/src/organisms/ProtocolTimelineScrubber/index.tsx:316]: make a const for `886`?

File: app/src/organisms/ProtocolTimelineScrubber/index.tsx [hunk 10]
```diff
@@ -0,0 +1,608 @@
...
+  const params: RunTimeCommand['params'] = command.params ?? {}
+  return (
+    <Flex
+      key={index}
+      backgroundColor={
+        index === currentCommandIndex
+          ? COLORS.blue35
+          : index < currentCommandIndex
+          ? '#00002222'
+          : '#fff'
```
Comment (koji) [inline app/src/organisms/ProtocolTimelineScrubber/index.tsx]:
```suggestion
          : COLORS.white
```

File: app/src/organisms/ProtocolTimelineScrubber/index.tsx [hunk 11]
```diff
@@ -0,0 +1,608 @@
...
+          x + 8
+        },170 L${x + 10},150 V130 H${x}z`}
+        stroke="#000"
+        fill="none"
+      />
+      <path
+        d={`M${x - 1},129 V150 L${x + 2},170 L${x + 4},180 H${x + 6} L${
+          x + 8
+        },170 L${x + 10},150 V130 H${x + 11} V181 H${x - 1} V129z`}
+        fill="#FFF"
```
Comment (koji) [inline app/src/organisms/ProtocolTimelineScrubber/index.tsx]:
```suggestion
        fill={COLORS.white}
```

File: app/src/organisms/ProtocolTimelineScrubber/index.tsx [hunk 12]
```diff
@@ -0,0 +1,608 @@
...
+    maxVolume -
+    Object.entries(tipContents).reduce((acc, [liquidId, { volume }]) => {
+      if (liquidId === '__air__') return acc
+      return acc + volume
+    }, 0)
+  const yOfFirstLiquid = (emptyVolumeLeft / maxVolume) * 50
+
+  return (
+    <>
+      <rect x={x} y={y} height={yOfFirstLiquid} width="10" fill="#FFF" />
```
Comment (koji) [inline app/src/organisms/ProtocolTimelineScrubber/index.tsx]:
```suggestion
      <rect x={x} y={y} height={yOfFirstLiquid} width="10" fill={COLORS.white} />
```

File: app/src/organisms/ProtocolTimelineScrubber/index.tsx [hunk 13]
```diff
@@ -0,0 +1,608 @@
...
+          ? `1px solid ${COLORS.blue35}`
+          : '1px solid #000'
+      }
+      padding={SPACING.spacing4}
+      flexDirection={DIRECTION_COLUMN}
+      minWidth={`${COMMAND_WIDTH_PX}px`}
+      width={`${COMMAND_WIDTH_PX}px`}
+      height="6rem"
+      overflowX="hidden"
+      overflowY="scroll"
```
Comment (koji) [inline app/src/organisms/ProtocolTimelineScrubber/index.tsx]:
```suggestion
      overflowX={OVERFLOW_HIDDEN}
      overflowY={OVERFLOW_SCROLL}
```

File: app/src/organisms/ProtocolTimelineScrubber/utils.ts [hunk 14]
```diff
@@ -0,0 +1,142 @@
...
+
+interface WellContentsByLabware {
+  [labwareId: string]: ContentsByWell
+}
+
+export type ContentsByWell = {
+  [wellName: string]: WellContents
+} | null
+
+const MIXED_WELL_COLOR = COLORS.grey50
```
Comment (koji) [inline app/src/organisms/ProtocolTimelineScrubber/utils.ts:34]: maybe just return `COLORS.grey50` in line 107?

File: app/src/organisms/ProtocolsLanding/ProtocolOverflowMenu.tsx [hunk 15]
```diff
@@ -108,6 +112,11 @@ export function ProtocolOverflowMenu(
     dispatch(analyzeProtocol(protocolKey))
     setShowOverflowMenu(currentShowOverflowMenu => !currentShowOverflowMenu)
   }
+  const handleClickTimeline: React.MouseEventHandler<HTMLButtonElement> = e => {
+    e.preventDefault()
+    history.push(`/protocols/${protocolKey}/timeline`)
+    setShowOverflowMenu(!showOverflowMenu)
```
Comment (koji) [inline app/src/organisms/ProtocolsLanding/ProtocolOverflowMenu.tsx]:
```suggestion
    setShowOverflowMenu(prevShowOverflowMenu => !prevShowOverflowMenu)
```

File: app/src/pages/Protocols/ProtocolDetails/ProtocolTimeline.tsx [hunk 16]
```diff
@@ -0,0 +1,36 @@
...
+  return storedProtocol != null && storedProtocol.mostRecentAnalysis != null ? (
+    <Box padding={SPACING.spacing16}>
+      <ProtocolTimelineScrubber
+        commands={storedProtocol.mostRecentAnalysis.commands}
+        analysis={storedProtocol.mostRecentAnalysis}
+        robotType={storedProtocol.mostRecentAnalysis.robotType}
+      />
+    </Box>
+  ) : (
+    <Icon size={SIZE_4} name="ot-spinner" spin />
```
Comment (koji) [inline app/src/pages/Protocols/ProtocolDetails/ProtocolTimeline.tsx]:
```suggestion
    <Icon size="8rem" name="ot-spinner" spin />
```

File: app/src/pages/Protocols/ProtocolDetails/ProtocolTimeline.tsx [hunk 17]
```diff
@@ -0,0 +1,36 @@
+import * as React from 'react'
+import { useParams } from 'react-router-dom'
+import { useDispatch, useSelector } from 'react-redux'
+import { Icon, SIZE_4, Box, SPACING } from '@opentrons/components'
```
Comment (koji) [inline app/src/pages/Protocols/ProtocolDetails/ProtocolTimeline.tsx]:
```suggestion
import { Icon, Box, SPACING } from '@opentrons/components'
```

File: app/src/organisms/ProtocolTimelineScrubber/index.tsx
Comment (jerader) [reply app/src/organisms/ProtocolTimelineScrubber/index.tsx:1]: good call, i'll separate them out!

File: app/src/App/DesktopApp.tsx
Comment (koji) [inline app/src/App/DesktopApp.tsx:1]: Can you add a test case of `ProtocolTimeline` to DesktopApp.test.tsx?
I think that wouldn't take so much time.

**Comment by koji** [comment]
@jerader 

Screenshot 2024-07-01 at 4 33 43 PM

I guess the space between deck map and timeline in the screenshot is smaller than the screen captures you posted.
Have you changed anything on that?

File: app/src/organisms/ProtocolTimelineScrubber/PipetteVisuals.tsx
Comment (koji) [inline app/src/organisms/ProtocolTimelineScrubber/PipetteVisuals.tsx:1]: If there any specific reason, using the same name for component name and .tsx file name.

**Comment by koji** [comment]
@jerader 

The error message came from `step-generation/src/getNextRobotStateAndWarnings/dispenseUpdateLiquidState.ts`.
It appears that you were the last person to edit the file. Do you have any idea to solve this error message or is this expected message?

Screenshot 2024-07-01 at 4 47 38 PM

**Comment by koji** [comment]
nit
maybe adding arrow icon (up/down) next to mount can be user-friendly. 

Screenshot 2024-07-01 at 4 55 25 PM

**Comment by koji** [comment]
warning

Screenshot 2024-07-01 at 4 59 52 PM

seems that `wait_for_duration` gets warning because of L361. `message` can be `undefined`.
if message == null, passing '' or something temporarily and leave a comment?

```ts
<LegacyStyledText as={as} {...styleProps}>
  {t('wait_for_duration', { seconds, message })}
</LegacyStyledText>
```

**Review by jerader — APPROVED** [review]

**Comment by jerader** [comment]
> nit maybe adding arrow icon (up/down) next to mount can be user-friendly.
> 
> Screenshot 2024-07-01 at 4 55 25 PM

@koji thanks for this! i'm gonna do a follow up pr to refine things a bit and will address this then.

**Review by koji — APPROVED** [review]
left a few comments.
the changes look good to me.

if you can issue tickets to follow up some tests, that will be helpful for us.

**Pull request merged**